lib/ecto/changeset/relation.ex

```diff
@@ -272,35 +278,38 @@ defmodule Ecto.Changeset.Relation do
   end
 
   defp current_or_nil(nil, _new_pks, current, _current_pks), do: current
-  defp current_or_nil(_new, _new_pks, nil, _current_pks), do: nil
   defp current_or_nil(new, new_pks, current, current_pks) do
     new_pk_values = get_pks(new, new_pks)
-    current_pk_values = get_pks(current, current_pks)
 
     cond do
-      new_pk_values == current_pk_values ->
+      current && new_pk_values == get_pks(current, current_pks) ->
         current
       Enum.all?(new_pk_values, &is_nil/1) ->
         nil
       true ->
-        raise Ecto.UnmachedRelationError, new_value: new, old_value: current, cardinality: :one
+        raise Ecto.UnmachedRelationError, new_value: new, old_value: current,
+                                          cardinality: :one
     end
   end
 
   defp get_pks(%Changeset{model: model}, pks),
     do: get_pks(model, pks)
   defp get_pks(model_or_params, pks),
-    do: Enum.map(pks, &Map.get(model_or_params, &1))
+    do: Enum.map(pks, &do_get_pk(model_or_params, &1))
 
-  defp primary_keys(module) do
-    primary_keys =
-      case module.__schema__(:primary_key) do
-        []  -> raise Ecto.NoPrimaryKeyFieldError, model: module
-        pks -> pks
-      end
-    param_keys = Enum.map(primary_keys, &Atom.to_string/1)
+  defp do_get_pk(model_or_params, {key, type}) do
+    case Ecto.Type.cast(type, do_get_pk(model_or_params, key)) do
+      {:ok, value} -> value
+      :error       -> :error
```

Comment:
This is a little bit hackish.
In case we can't cast a key we'll get `[:error]` instead of `[1]` or something, and that obviously won't match any id or binary_id, but I'm not sure it's ok to do that.